pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go

```diff
@@ -999,25 +1060,65 @@ func (lbaas *LbaasV2) EnsureLoadBalancer(clusterName string, apiService *v1.Serv
 		}
 
 		_, err = rules.Create(lbaas.network, lbSecGroupRuleCreateOpts).Extract()
-
 		if err != nil {
-			// cleanup what was created so far
-			_ = lbaas.EnsureLoadBalancerDeleted(clusterName, apiService)
-			return nil, err
+			return fmt.Errorf("Error occured creating rule for SecGroup %s: %v", lbSecGroup.ID, err)
 		}
 
+		// get security groups of port
 		portID := loadbalancer.VipPortID
-		update_opts := neutronports.UpdateOpts{SecurityGroups: &[]string{lbSecGroup.ID}}
-		res := neutronports.Update(lbaas.network, portID, update_opts)
-		if res.Err != nil {
-			glog.Errorf("Error occured updating port: %s", portID)
-			// cleanup what was created so far
-			_ = lbaas.EnsureLoadBalancerDeleted(clusterName, apiService)
-			return nil, res.Err
+		port, err := getPortByID(lbaas.network, portID)
+		if err != nil {
+			return err
+		}
+
+		// ensure the vip port has the security groups
+		found := false
+		for _, portSecurityGroups := range port.SecurityGroups {
+			if portSecurityGroups == lbSecGroup.ID {
+				found = true
+				break
+			}
+		}
+
+		// update loadbalancer vip port
+		if !found {
+			port.SecurityGroups = append(port.SecurityGroups, lbSecGroup.ID)
+			update_opts := neutronports.UpdateOpts{SecurityGroups: &port.SecurityGroups}
+			res := neutronports.Update(lbaas.network, portID, update_opts)
+			if res.Err != nil {
+				msg := fmt.Sprintf("Error occured updating port %s for loadbalancer service %s/%s: %v", portID, apiService.Namespace, apiService.Name, res.Err)
+				glog.Errorf(msg)
+				return fmt.Errorf(msg)
+			}
 		}
 	}
 
-	return status, nil
+	// ensure rules for every node security group
+	for _, port := range ports {
+		for _, nodeSecurityGroupID := range lbaas.opts.NodeSecurityGroupIDs {
+			opts := rules.ListOpts{
+				SecGroupID:    nodeSecurityGroupID,
+				RemoteGroupID: lbSecGroupID,
+			}
+			secGroupRules, err := getSecurityGroupRules(lbaas.network, opts)
+			if err != nil && !isNotFound(err) {
+				glog.Errorf("Error finding rules for remote group id %s in security group id %s", lbSecGroupID, nodeSecurityGroupID)
+				return err
+			}
+			if len(secGroupRules) != 0 {
+				// Do not add rule when find rules for remote group in the Node Security Group
+				continue
+			}
+
+			// Add the rules in the Node Security Group
+			err = createNodeSecurityGroup(lbaas.network, nodeSecurityGroupID, int(port.NodePort), port.Protocol, lbSecGroupID)
+			if err != nil {
+				return fmt.Errorf("Error occured creating security group for loadbalancer service %s/%s", apiService.Namespace, apiService.Name)
```

Comment:
Always include `err` in the returned error, or someone looking at the logs will have to guess what actually went wrong.